Add support for itemset member tracking when input transactions contain IDs

CRJFisher opened the pull request:
This allows the client to pass through transactions with IDs, signalling that the algorithm should keep track of _where_ the frequent itemsets occur. 
The results then contain itemsets with their counts as well as the IDs of transactions where they occurred.

The main conceptual changes were in itemsets.py:

1. TransactionBroker - handles fetching rows of transactions for all the different types of input.
2. CountBroker - handles counting for all transaction input types. For transactions with ids, the count method also keeps track of the transaction ID. The idea here was to keep the core of the algorithm code semantically the same.

The new test file might look a bit familiar since it's copied from test_itemsets.py but with changes to cover all the ways IDs can be passed in.

The original code was a joy to work with - hopefully I haven't changed that too much :)

tommyod commented:
Thanks for the contribution. I'll review it more carefully soon. Some questions/comments:
- Currently "flake8" fails. Can you make sure all tests run on travis pass?
- Could you write a paragraph about the use case of this?
- How does the computational complexity compare with (1)getting itemssets are normally done with the algorithm, (2) going through transactions and (3) checking if each itemset is a subset of the transaction?
- If the computational complexity is much better than the naive approach outlined above, would it be an idea to instead just return the indices of the itemsets? Then the algorithm could run as normal, and you could query and get indices 1, 2, ... , N. Now the user must deal with a whole new type of object, as well as manually assigning IDs before feeding it to the algorithm. I see no loss of generality in assuming IDs 1, 2, ..., N, and then the user would not have to deal with custom types.

Thoughts on the above?

I will review more carefully, those are just my preliminary thoughts. Thanks again! :+1:

CRJFisher commented:
>     * Currently "flake8" fails. Can you make sure all tests run on travis pass?
That should be all fixed now.

>     * Could you write a paragraph about the use case of this?
The use cases are when downstream analysis is required where knowing which transactions occurred in a frequent itemset is useful. There could be some other data-point associated with the transaction that is interesting to analyse for a given frequent itemset.
For example, if the transactions are items bought in shops, there might be some geographical or time-based information which let you answer where or when were the shopping itemsets frequent.

>     * How does the computational complexity compare with (1)getting itemssets are normally done with the algorithm, (2) going through transactions and (3) checking if each itemset is a subset of the transaction?
These are all identical to apriori v1.0.0.
In the original PR commits, there was an extra function call in the main transaction loops which would potentially add some overhead; however, in the latest changes, that has been addressed and removed for the non-id-tracking (original) use case. 
This was done by putting the counter-specific functionality, including the main transaction scanning loops into two classes sharing a single interface (`_Counter` and `_CounterWithIds`). This meant that in the `_Counter` implementation, the count can be incremented in the original way (`counts[candidate] += 1`) without the need of the function call used in the id-tracking variant (`counts[candidate].increment_count(row)`).

>     * Why not simply assume IDs 1, 2, ... , N? Now the user must deal with a whole new type of object, as well as manually assigning IDs before feeding it to the algorithm. I see no loss of generality in assuming IDs 1, 2, ..., N, and then the user would not have to deal with custom types. The user could give a list of transactions as before, they would implicitly be given IDs 1, 2, ..., N, and those IDs (or corresponding transactions) could be returned.
I see what you mean here. I've removed the TransactionWithId class and instead, added a `output_transaction_ids` flag to signal the id-tracking behaviour. The ids are then the enumeration of the transactions in the order they appear. The main benefit of passing bespoke ids was in case the client had some pre-existing transaction index (as I do in my use-case!) but that can be easily handled by some mapping on the client side.

tommyod commented:
Thank you very much @CRJFisher . This is a nice PR in all respects, and I'll happily merge.

Great job!
